tests/util/caches/test_descriptors.py

```diff
@@ -478,10 +446,10 @@ async def func1(self, key, cache_context):
 
             @cached(cache_context=True)
             async def func2(self, key, cache_context):
-                return self.func3(key, on_invalidate=cache_context.invalidate)
+                return await self.func3(key, on_invalidate=cache_context.invalidate)
```

Comment:
I am a bit confused why these changes are needed, given we seem to use `maybeDeferred`:

https://github.com/matrix-org/synapse/blob/19c0e55ef7742d67cff1cb6fb7c3e862b86ea788/synapse/util/caches/descriptors.py

Reply:
I would love to understand it more. Traceback for completeness.

I spent some time staring at this in a debugger and couldn't divine any insight.

```
==============================================================================
Error: 
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/util/caches/test_descriptors.py", line 459, in test_invalidate_cascade
    r = get_awaitable_result(obj.func1("k1", on_invalidate=top_invalidate))
  File "/home/runner/work/synapse/synapse/tests/test_utils/__init__.py", line 40, in get_awaitable_result
    next(i)
  File "/home/runner/work/synapse/synapse/tests/util/caches/test_descriptors.py", line 445, in func1
    return await self.func2(key, on_invalidate=cache_context.invalidate)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/runner/work/synapse/synapse/synapse/logging/context.py", line 753, in g
    return run_in_background(f, *args, **kwargs)
  File "/home/runner/work/synapse/synapse/synapse/logging/context.py", line 814, in run_in_background
    res = defer.ensureDeferred(res)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 911, in ensureDeferred
    return _cancellableInlineCallbacks(coro)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 1421, in _inlineCallbacks
    status.deferred.callback(getattr(e, "value", None))
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.7/lib/python3.7/site-packages/twisted/internet/defer.py", line 459, in callback
    assert not isinstance(result, Deferred)
builtins.AssertionError: 

tests.util.caches.test_descriptors.DescriptorTestCase.test_invalidate_cascade
-------------------------------------------------------------------------------
```